internal/app/secretless/handlers/mysql/authentication_handshake.go

```diff
@@ -0,0 +1,380 @@
+	h.handleClientSSLRequest()
+	h.writeClientHandshakeResponseToBackend()
+
+	// TODO: Kumbi: what does this mean? AuthSwitchRequest, AuthSwitchResponse
+	// https://dev.mysql.com/doc/refman/8.0/en/authentication-plugins.html
+
+	h.verifyAndProxyOkResponse()
+	return h.err
+}
+
+// AuthenticatedBackendConn returns an already authenticated connection
+// to the MySQL server.  Intended to be called after Run() has completed.
+//
+func (h *AuthenticationHandshake) AuthenticatedBackendConn() net.Conn {
+	return h.backendConn.RawConnection()
+}
+
+func (h *AuthenticationHandshake) readServerHandshake() {
+	if h.err != nil {
+		return
+	}
+	h.rawServerHandshake = h.readBackendPacket()
+
+	if h.err != nil {
+		return
+	}
+	h.serverHandshake, h.err = protocol.UnpackHandshakeV10(h.rawServerHandshake)
+}
+
+func (h *AuthenticationHandshake) writeHandshakeToClient() {
+	if h.err != nil {
+		return
+	}
+
+	// Remove Client SSL Capability from Server Handshake Packet
+	// to force client to connect to Secretless without SSL
+	// TODO: update this after kumbi's work
+	//
+	packetWithNoSSL, err := protocol.RemoveSSLFromHandshakeV10(h.rawServerHandshake)
+	if err != nil {
+		h.err = err
+		return
+	}
+
+	// TODO: push all packing code into the `WriteXXX` methods
+	h.writeClientPacket(packetWithNoSSL)
+}
+
+func (h *AuthenticationHandshake) validateServerSSL() {
+	if h.err != nil {
+		return
+	}
+
+	if h.clientRequestedSSL() && !h.serverSupportsSSL() {
+		h.err = errNoTLS
+	}
+}
+
+func (h *AuthenticationHandshake) readClientHandshakeResponse() {
+	if h.err != nil {
+		return
+	}
+
+	rawResponse := h.readClientPacket()
+	if h.err != nil {
+		return
+	}
+
+	// TODO: client requesting SSL results ERROR 2026 (HY000): SSL connection error: protocol version mismatch
+	h.clientHandshakeResponse, h.err = protocol.UnpackHandshakeResponse41(rawResponse)
+}
+func (h *AuthenticationHandshake) overrideClientCapabilities() {
+	if h.err != nil {
+		return
+	}
+
+	// TODO: after kumbi's done, change below to method calls
+
+	// TODO: add tests cases for authentication plugins support
+	// Disable CapabilityFlag for authentication plugins support
+	h.clientHandshakeResponse.CapabilityFlags &^= protocol.ClientPluginAuth
+
+	// TODO: add tests cases for client secure connection
+	// Enable CapabilityFlag for client secure connection
+	h.clientHandshakeResponse.CapabilityFlags |= protocol.ClientSecureConnection
+
+	// Ensure CapabilityFlag is set when using TLS
+	if h.clientRequestedSSL() {
+		h.clientHandshakeResponse.CapabilityFlags |= protocol.ClientSSL
+	}
+
+}
+
+func (h *AuthenticationHandshake) injectCredentials() {
+	if h.err != nil {
+		return
+	}
+
+	// TODO: change this to method call on clientHandshakeResponse when Kumbi's work done
+	h.err = protocol.InjectCredentials(
+		h.clientHandshakeResponse,
+		h.serverHandshake.Salt,
+		h.connectionDetails.Username,
+		h.connectionDetails.Password,
+	)
+}
+
+// TODO: All of this should be moved to kumbi's code
```

Comment:
just to verify, that's within the protocol? because if we're getting into manipulating packets, I think that's where it belongs.

Reply:
Removed, although this code will likely change when we do the ssl utility code changes